Save custom decks to the database

greg19 opened the pull request:
Fixes #735.

greg19 committed:
Deck builder changes

greg19 committed:
Deck public status

greg19 commented:
@marcelpanse can you create the DB table? I think this will be okay:

```sql
CREATE TABLE decks (
  id SERIAL PRIMARY KEY,
  email VARCHAR(?) NOT NULL REFERENCES accounts(email) ON UPDATE CASCADE,
  is_public BOOLEAN NOT NULL,
  name VARCHAR(64) NOT NULL CHECK (length(name) > 0),
  energy VARCHAR(24)[] NOT NULL CHECK (array_length(energy, 1) > 0),
  cards INTEGER[] NOT NULL
);
```

Unrelated note: Can you verify that `card_amounts` table has `email` foreign key set? I can't see it

marcelpanse commented:
Table is created. 

Seems there are no FKs to the accounts table. Not sure why they were never created, best guess is because I started initially on Appwrite which didn't really have FKs and later migrated to Supabase.

greg19 committed:
wip

greg19 committed:
wip

greg19 committed:
toasts

greg19 committed:
spinner

greg19 committed:
wip

greg19 committed:
wip

greg19 committed:
Merge main

greg19 committed:
Tabs and prettier list

greg19 committed:
Prepare public decks

greg19 commented:
@marcelpanse can you also create a `public_decks` view?

greg19 committed:
Public decks

greg19 committed:
Merge remote-tracking branch 'origin/main' into save-decks

greg19 committed:
Error alert

greg19 commented:
@marcelpanse can you also do:

```sql
CREATE TABLE deck_likes (
  id INTEGER NOT NULL REFERENCES decks(id),
  email VARCHAR(?) NOT NULL REFERENCES accounts(email) ON UPDATE CASCADE,
  PRIMARY KEY (deck_id, email)
);
```

This will serve two purposes. The first is bookmarking a deck, I want to also shown liked decks on "my decks" tab (or have 4th tab with the liked decks).

The second purpose is searching for decks. The "community decks" would have the decks sorted by the number of likes, so you won't be shown junk (that will inevitably appear if the users are allowed to post arbitrary decks). For this, we need to somehow count the likes. Counting them every time someone fetches the table isn't scalable, so we need to do some bookkeeping in the db.

```sql
CREATE TABLE deck_like_counts (
  id INTEGER PRIMARY KEY REFERENCES decks(id),
  likes INTEGER NOT NULL DEFAULT 0
);

CREATE INDEX deck_like_counts_order_idx
  ON deck_like_counts (likes DESC)
  INCLUDE (deck_id);
```

We can't give write access for users for this table. So liking and disliking a deck either has to be a RPC / deno call, or we need the triggers:

```sql
REVOKE ALL ON deck_like_counts FROM PUBLIC;
GRANT SELECT ON deck_like_counts TO PUBLIC;

CREATE OR REPLACE FUNCTION deck_likes_after_insert()
RETURNS TRIGGER
LANGUAGE plpgsql
SECURITY DEFINER
SET search_path = pg_catalog
AS $$
BEGIN
  INSERT INTO public.deck_like_counts (deck_id, likes)
  VALUES (NEW.deck_id, 1)
  ON CONFLICT (deck_id)
  DO UPDATE SET likes = deck_like_counts.likes + 1;

  RETURN NULL;
END;
$$;

CREATE OR REPLACE FUNCTION deck_likes_after_delete()
RETURNS TRIGGER
LANGUAGE plpgsql
SECURITY DEFINER
SET search_path = pg_catalog
AS $$
BEGIN
  UPDATE public.deck_like_counts
     SET likes = GREATEST(likes - 1, 0)
   WHERE deck_id = OLD.deck_id;

  RETURN NULL;
END;
$$;

CREATE TRIGGER trg_deck_likes_after_insert
AFTER INSERT ON deck_likes
FOR EACH ROW
EXECUTE FUNCTION deck_likes_after_insert();

CREATE TRIGGER trg_deck_likes_after_delete
AFTER DELETE ON deck_likes
FOR EACH ROW
EXECUTE FUNCTION deck_likes_after_delete();
```

greg19 committed:
Deck view page

greg19 committed:
Improve like button

greg19 committed:
Liked decks page

marcelpanse commented:
I created the decks_likes table and I added the likes in the public_decks view using a join. So doing `select * from public_decks` will give you a column `likes` that you can sort on. 

I think performance-wise this will be fine. I don't expect that many decks being made and liked, so that shouldn't be a problem at all. 

If it does become super-popular somehow and the query becomes slow, then the best way to solve it is to make another cron job like the `stats-tracker` function that runs once per hour, does the query, and stores the results to a static json file. Then the app will load the static file to show the popular decks. This approach is super scalable and doesn't add any complexity to the database. The added benefit is that the like_count can never be out of sync with the actual deck_like table since it works on the actual data instead of some increments.

greg19 committed:
wip

greg19 committed:
wip

greg19 committed:
wip

greg19 committed:
filter by energy

greg19 committed:
responsive view

greg19 committed:
order

greg19 committed:
Improve ui

greg19 committed:
Override bugged deckbuilding promos

greg19 committed:
Adjust rarity filter for deckbuilding mode

greg19 committed:
Fix saving deck

greg19 committed:
Validate decks with zod

greg19 committed:
Validate decks with zod

greg19 committed:
Remove commented out code

greg19 committed:
Fix on delete invalidate

greg19 committed:
New deck name

Sets the default name of a new deck to an empty string to motivate the
user to set the name to something different than "New deck"

greg19 committed:
Deck name placeholder

marcelpanse commented:
Regarding the scraper, maybe we can adjust it to import the meta decks into the database (like in my account for example). We can also see how this will be used, maybe some people will start adding the meta decks themselves. So lets keep the scraper file for now and decide later.

greg19 commented on frontend/src/pages/decks/DeckBuilder.tsx:
```diff
@@ -27,16 +35,43 @@ const defaultFilters: Filters = {
+          setIsError(true)
+        })
+    }
+  }, [shouldFetch, deckId])
```
Because you need to be able to edit the deck, which itself requires a `useState`. Then `useEffect` is simpler than trying to put the `useDeck` here (`useQuery` doesn't produce any event when it fetches the data, so you have no place to call the initial `setDeck` after fetching form the db)

greg19 commented:
Okay, as you probably seen I've created some decks. Though deck id=8 is behaving weirdly, it no longer shows in my decks, and when I try to like it I get an error `new row violates row-level security policy for table "deck_likes"`. Also, can you give me the read access in postgres to the newly created tables?

greg19 committed:
Fix deck copy

greg19 commented:
> when I now try to copy and edit your deck, it will fail because it seems it tries to edit the deck instead of making a new copy of it. Can you fix that?

Should be fixed

marcelpanse commented:
> I now don't have access to any of the tables

you should have access now to all tables

greg19 committed:
Fix on update invalidate

greg19 commented:
Fixed
